### PR TITLE
Many symintifications

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87722
* #87294
* __->__ #87604

Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes

### CODE REVIEW

**Comment by albanD** [comment]
Will wait for CI for all the xfail update and make sure everything is tested.

File: aten/src/ATen/native/NonSymbolicBC.h [hunk 1]
```diff
@@ -22,4 +22,5 @@ TORCH_API at::Tensor _embedding_bag_sparse_backward(const at::Tensor & grad, con
 TORCH_API at::Tensor value_selecting_reduction_backward(const at::Tensor & grad, int64_t dim, const at::Tensor & indices, at::IntArrayRef sizes, bool keepdim);
 TORCH_API at::Tensor trace_backward(const at::Tensor & grad, at::IntArrayRef sizes);
 TORCH_API at::Tensor index_select_backward(const at::Tensor & grad, at::IntArrayRef self_sizes, int64_t dim, const at::Tensor & index);
+TORCH_API std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_t dim);
```
Comment (bdhirsh) [inline aten/src/ATen/native/NonSymbolicBC.h:25]: This has a small risk of internal breakage, right?

Although I just searched internal for `at::native::tensor_split`, and the only place that shows up is in static runtime (which is part of the OSS codebase). So.. I think this should be fine

File: aten/src/ATen/native/Pool.h [hunk 2]
```diff
@@ -67,24 +67,34 @@ static inline T pooling_output_shape(
...
-  auto total_padding = dilation * (kernelSize - 1);
+  auto total_padding = T(dilation) * (kernelSize - 1);
 
   // Prefer symmetric padding if possible
   if (stride > 2 && (total_padding % 2 == 1)) {
     // The floor in the output size calculation gives us a little wiggle room
     auto wiggle_room = inputSize % stride - 1;
     if (wiggle_room > 0) {
-      --total_padding;
+      total_padding = total_padding - 1;
```
Comment (bdhirsh) [inline aten/src/ATen/native/Pool.h:81]: Do you think it's worth implementing `operator--` as part of merging to master? (I'll leave it up to you)

Comment (ezyang) [reply aten/src/ATen/native/Pool.h:81]: seems worthwhile. But I'll take a rain check if someone agrees to do ALL of the operator overloads, methodically

File: aten/src/ATen/native/TensorShape.cpp [hunk 3]
```diff
@@ -756,21 +759,30 @@ std::vector<Tensor> tensor_split(const Tensor& self, int64_t sections, int64_t d
...
+  T start_idx(0);
   for (const auto split_idx : c10::irange(num_indices)) {
-    int64_t end_idx = indices[split_idx];
-    splits[split_idx] = at::slice(self, dim_, start_idx, end_idx);
+    auto end_idx = indices[split_idx];
+    splits[split_idx] = at::symint::slice<T>(self, dim_, start_idx, end_idx);
     start_idx = end_idx;
   }
-  splits[num_indices] = at::slice(self, dim_, start_idx, self.size(dim_));
+  splits[num_indices] = at::symint::slice<T>(self, dim_, start_idx, self.size(dim_));
```
Comment (bdhirsh) [inline aten/src/ATen/native/TensorShape.cpp:774]: hmmm don't we need to symintify that `self.size(dim_)`?

I guess we could add a template for `at::symint::size()`, and make it:
```
at::symint::size<T>(self)`
```
wdyt?

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -1815,7 +1821,7 @@
 - func: diag_embed(Tensor self, int offset=0, int dim1=-2, int dim2=-1) -> Tensor
   variants: function, method
   dispatch:
-    CompositeExplicitAutogradNonFunctional: diag_embed
+    CompositeExplicitAutograd: diag_embed
```
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:1824]: hmmm, why?

`CompositExplicitAutogradNonFunctional` is a bit funny today - it's basically only used for providing better error messages for XLA/LTC. If we have a functional op that calls a view op, that is `CompositeExplicitAutograd`, then XLA and LTC can't handle it directly so we don't want to register it to them. Not clear to me why we had to remove it for `diag_embed` here though.

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -7698,10 +7704,21 @@
 
 - func: diag.out(Tensor self, int diagonal=0, *, Tensor(a!) out) -> Tensor(a!)
   dispatch:
-    CPU, CUDA: diag_out
+    CPU: diag_cpu_out
+    CUDA: diag_cuda_out
+    MPS: diag_mps_out
 
 - func: diag(Tensor self, int diagonal=0) -> Tensor
```
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:7711]: Is there a comment explaining the reasoning for the change to `diag`?

It looks like before, `diag` was a `CompositeImplicitAutograd` op that called `diag_embed`, and now we have a brand new backward op for it.

File: aten/src/ATen/native/TensorShape.cpp [hunk 6]
```diff
@@ -3666,16 +3721,8 @@ at::Tensor& _sparse_broadcast_to_copy_out(const at::Tensor & self, at::IntArrayR
...
-  TORCH_CHECK(
-    out.device() == self.device(),
-    "diagonal_copy: Expected out and self tensors to be on the same device, but got ",
-    "out on ", out.device(), " and self on ", self.device());
-  auto result = self.diagonal(offset, dim1, dim2);
-  at::native::resize_output(out, result.sizes());
-  TORCH_CHECK(
-      canCast(result.scalar_type(), out.scalar_type()),
-      "diagonal_copy: result type ", result.scalar_type(), " can't be cast to the desired out= type ", out.scalar_type());
-  out.copy_(result);
```
Comment (bdhirsh) [inline aten/src/ATen/native/TensorShape.cpp:3678]: This seems wrong - we're removing error checking?

File: aten/src/ATen/native/NonSymbolicBC.h (hunk 1 above)
Comment (albanD) [reply aten/src/ATen/native/NonSymbolicBC.h:25]: Ok, let's take the risk on this one. Thanks for checking!

File: aten/src/ATen/native/TensorShape.cpp (hunk 3 above)
Comment (albanD) [reply aten/src/ATen/native/TensorShape.cpp:774]: Yes I can add that.

File: aten/src/ATen/native/native_functions.yaml (hunk 4 above)
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:1824]: Not sure tbh, I haven't reviewed this. Reverting.

File: torchgen/api/cpp.py [hunk 7]
```diff
@@ -342,11 +342,13 @@ def default_expr(d: str, t: Type) -> str:
...
             return "c10::nullopt"
 
-        return default_expr(d, t.elem)
+        return default_expr(d, t.elem, symint=symint)
 
     if isinstance(t, ListType):
         if d.startswith("[") and d.endswith("]"):
             return "{" + d[1:-1] + "}"
+        elif symint and d.isdigit() and str(t.elem) == "SymInt":
+            return f"c10::SymInt({d})"
```
[the remark this answers is not in the document]
Comment (albanD) [reply torchgen/api/cpp.py:351]: It is actually used for `SymInt[3] padding=0` so it is for the list case.

File: aten/src/ATen/native/TensorShape.cpp [hunk 8]
```diff
@@ -3390,29 +3402,72 @@ Tensor unfold(const Tensor& self, int64_t d, int64_t size, int64_t step) {
   return self.as_strided(sizes, strides);
 }
 
-Tensor diag(const Tensor& self, int64_t offset) {
-  auto ndim = self.dim();
-  TORCH_CHECK(ndim == 1 || ndim == 2, "diag(): Supports 1D or 2D tensors. Got ", self.dim(), "D");
-  if (ndim == 1) {
-    return at::diag_embed(self, offset);
+template <typename scalar_t>
+void apply_diag(Tensor& result, const Tensor& self, int64_t dimension) {
```
Comment (ezyang) [inline aten/src/ATen/native/TensorShape.cpp:3406]: The diff here is weirdly big and makes me nervous

**Comment by ezyang** [comment]
Everything here looks good except diag, which is weirdly complicated for what I expect

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:7711]: This is a mistake. The diag improvements were added to master indepently and never added to the branch.
I'll make sure not to revert https://github.com/pytorch/pytorch/pull/87180 by mistake :p

**Comment by albanD** [comment]
>  Everything here looks good except diag, which is weirdly complicated for what I expect

Yes the diag change is a mistake due to a land race with master. Cleaning this up now.

File: aten/src/ATen/native/Pool.h (hunk 2 above)
Comment (albanD) [reply aten/src/ATen/native/Pool.h:81]: Ok, leaving all the operators for a follow up then.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**